src/box/memtx_engine.cc

```diff
@@ -1006,13 +1014,16 @@ checkpoint_f(va_list ap)
 		ckpt->touch = false;
 	}
 
-	struct xlog snap;
-	if (xdir_create_xlog(&ckpt->dir, &snap, &ckpt->vclock) != 0)
+	struct xlog *snap = &ckpt->snap;
+	assert(!xlog_is_open(snap));
+	if (xdir_create_xlog(&ckpt->dir, snap, &ckpt->vclock) != 0) {
+		xlog_clear(snap);
```

Comment:
Makes sense. Done.

```diff
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 2f22a1232edd..1d1ce6bffc92 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -1017,6 +1017,11 @@ checkpoint_f(va_list ap)
 	struct xlog *snap = &ckpt->snap;
 	assert(!xlog_is_open(snap));
 	if (xdir_create_xlog(&ckpt->dir, snap, &ckpt->vclock) != 0) {
+		/*
+		 * We call memtx_engine_abort_checkpoint on failure to discard
+		 * an incomplete xlog file. Clear the xlog object so that it's
+		 * safe to call xlog_discard() on it.
+		 */
 		xlog_clear(snap);
 		return -1;
 	}
```